File: cli/lib/methods/apertium.js
/**
 * Apertium Translation Method — free/open-source RULE-BASED MT via the public
 * Apertium APy REST API. Mirrors the harness ApertiumMethod
 * (arena/mt_eval_harness/methods/apertium.py): same endpoint, env vars, and
 * GET/responseData shape, so config is interchangeable between CLI and harness.
 *
 *   Endpoint: GET <base>/translate?langpair=<src>|<tgt>&q=<text>&markUnknown=no
 *             default https://apertium.org/apy
 *   Auth:     none (public). Keyed APy deployments accept ?key=<key>.
 *   Env:      APERTIUM_API_URL (optional), APERTIUM_API_KEY (optional)
 *   Locale:   pass-through (accepts ISO 639-1 'es' or 639-3 'spa')
 *   License:  GPL — called as a separate service, never bundled.
 *
 * Only installed pairs work (strongest on related languages). An uninstalled
 * pair returns HTTP 400 and is surfaced honestly, never faked.
 */

import { TranslationMethod } from './base.js';
import { getEnvOrFileVar } from '../api-key.js';
import { DEFAULT_METHOD_CONCURRENCY } from '../config.js';
import { extractContentBody } from './content-separator.js';
import { output } from '../output.js';
import { fetchWithRetry } from './fetch-with-retry.js';
import { pMap } from '../concurrent.js';

const APERTIUM_DEFAULT_BASE = 'https://apertium.org/apy';
const APERTIUM_USER_AGENT =
  'Champollion-MT-Eval/0.1 (+https://champollion.dev; champollion CLI)';
const APERTIUM_TIMEOUT_MS = 15000;

class ApertiumMethod extends TranslationMethod {
  constructor(options = {}) {
    super('apertium', options);
  }

  _resolveEndpoint(options = {}) {
    let ep = options.apertiumApiUrl
      || getEnvOrFileVar('APERTIUM_API_URL')
      || getEnvOrFileVar('APERTIUM_API_URL', options.cwd)
      || APERTIUM_DEFAULT_BASE;
    ep = ep.replace(/\/+$/, '');
    if (ep.endsWith('/translate')) ep = ep.slice(0, -'/translate'.length);
    return ep;
  }

  _resolveApiKey(options = {}) {
    return options.apertiumApiKey
      || getEnvOrFileVar('APERTIUM_API_KEY')
      || getEnvOrFileVar('APERTIUM_API_KEY', options.cwd);
  }

  _buildUrl(base, langpair, text, apiKey) {
    const params = new URLSearchParams({ langpair, q: text, markUnknown: 'no' });
    if (apiKey) params.set('key', apiKey);
    return `${base}/translate?${params.toString()}`;
  }

  async _translateOne(base, langpair, text, apiKey, label) {
    const res = await fetchWithRetry(
      this._buildUrl(base, langpair, text, apiKey),
      { method: 'GET', headers: { 'User-Agent': APERTIUM_USER_AGENT } },
      { label, timeoutMs: APERTIUM_TIMEOUT_MS },
    );
    if (!res) return null;
    if (!res.ok) {
      const body = await res.text();
      output.error(`Apertium ${label}: ${res.status} — ${body.slice(0, 200)}`);
      return null;
    }
    const json = await res.json();
    const translated = json?.responseData?.translatedText;
    return typeof translated === 'string' ? translated : null;
  }

  async translate(keys, sourceFlat, pairConfig, options) {
    const base = this._resolveEndpoint(options);
    const apiKey = this._resolveApiKey(options);
    const langpair = `${pairConfig.source || 'en'}|${pairConfig.target}`;
    const allTranslated = {};

    const items = keys.filter(
      (k) => typeof sourceFlat[k] === 'string' && sourceFlat[k],
    );
    await pMap(items, async (key) => {
      const out = await this._translateOne(base, langpair, sourceFlat[key], apiKey, `key ${key}`);
      if (out !== null) allTranslated[key] = out;
    }, { concurrency: DEFAULT_METHOD_CONCURRENCY });

    return Object.keys(allTranslated).length > 0 ? allTranslated : null;
  }

  async translateContent(prompt, pairConfig, options) {
    const base = this._resolveEndpoint(options);
    const apiKey = this._resolveApiKey(options);
    const bodyText = extractContentBody(prompt);
    if (!bodyText.trim()) return null;
    const langpair = `${pairConfig.source || 'en'}|${pairConfig.target}`;
    return this._translateOne(base, langpair, bodyText, apiKey, 'content');
  }

  async checkReadiness(context) {
    const base = this._resolveEndpoint(context || {});
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 5000);
      const res = await fetch(`${base}/listPairs`, {
        method: 'GET',
        headers: { 'User-Agent': APERTIUM_USER_AGENT },
        signal: controller.signal,
      });
      clearTimeout(timeoutId);
      if (!res.ok) {
        return { ready: false, reason: `Apertium API at ${base} responded ${res.status}.` };
      }
      return { ready: true };
    } catch (err) {
      return {
        ready: false,
        reason:
          `Cannot reach Apertium API at ${base}: ` +
          `${err.name === 'AbortError' ? 'timeout' : err.message}. ` +
          `Override with APERTIUM_API_URL or self-host APy.`,
      };
    }
  }

  estimateCost(_keyCount) {
    // The public Apertium API is free (rate-limited) — there is no per-character
    // charge, so $0 is honest here (unlike local LLMs, whose compute cost is
    // genuinely unknown).
    return {
      estimatedCost: 0,
      currency: 'USD',
      source: 'apertium-free-public-api',
      note: 'Apertium public API is free (rate-limited; self-host for volume).',
    };
  }

  getQualityTier() {
    return 'standard';
  }

  getProvenance() {
    return {
      resources: [
        { name: 'Apertium (rule-based MT)', license: 'GPL-3.0+', type: 'api' },
      ],
      commercialReady: false,
      flags: ['rule-based', 'gpl'],
    };
  }

  getSetupHelp() {
    return [
      '  Apertium (rule-based, free public API):',
      '   • No API key needed — uses https://apertium.org/apy by default.',
      '   • Only installed pairs work (strong on related languages, e.g. es↔ca, eng↔spa).',
      '   • Self-host or point elsewhere: export APERTIUM_API_URL=https://your-apy',
    ];
  }
}

export { ApertiumMethod };

File: arena/mt_eval_harness/methods/apertium.py
"""
ApertiumMethod — Apertium public REST API (APy) adapter.

Apertium is free/open-source RULE-BASED (shallow-transfer) MT — it populates
the ``rule-based`` cell of the leaderboard's paradigm axis (everything else so
far is neural-nmt or llm). It is strongest on closely-related language pairs
(es↔ca, es↔pt, eng↔cat, …). Hosted as a separate GPL service; we only call its
public API — no Apertium code is bundled (license boundary preserved).

    Endpoint:  GET <base>/translate?langpair=<src>|<tgt>&q=<text>&markUnknown=no
               default https://apertium.org/apy
    Auth:      none (public API). Keyed APy deployments accept ?key=<key>.
    Env:       APERTIUM_API_URL (optional endpoint override),
               APERTIUM_API_KEY (optional — only for keyed deployments)
    Batch:     1 segment / request (the translate endpoint takes one q)
    Response:  json["responseData"]["translatedText"]
    Locale:    pass-through — Apertium accepts ISO 639-1 ("es") or 639-3 ("spa")
    License:   GPL (Apertium) — invoked as a separate service, never bundled
    Cost:      free public API; rate-limited — be polite, self-host for volume

NOTE: not every pair is installed on the public instance (e.g. eng→fra returns
HTTP 400 "That pair is not installed"). That surfaces as a normal per-entry
error via the base translate() loop — honest, never faked.
"""

from __future__ import annotations

import aiohttp

from mt_eval_harness.methods.base_http_mt import HttpMTMethod, _env_first

APERTIUM_DEFAULT_BASE = "https://apertium.org/apy"

# Identifying User-Agent so Apertium sees attributable, creditable traffic
# (design directive: always attribute corpus/tool providers).
APERTIUM_USER_AGENT = (
    "Champollion-MT-Eval/0.1 (+https://champollion.dev; MT eval harness)"
)


class ApertiumMethod(HttpMTMethod):
    """Apertium rule-based MT system-under-test (public APy)."""

    name = "apertium"
    MAX_BATCH = 1  # the APy translate endpoint takes a single q text

    method_id = "apertium"
    # method_class = "api" (inherited). The headline: this is the rule-based
    # paradigm — the empty cell the method axis was built to fill.
    paradigm = "rule-based"
    author = "Apertium (open source)"
    description = (
        "Apertium — free/open-source rule-based (shallow-transfer) MT; "
        "strongest on closely-related language pairs."
    )
    homepage = "https://www.apertium.org"
    license = "GPL-3.0+ (Apertium)"
    # GPL service invoked over HTTP (never bundled). Flag for license review
    # rather than asserting commercial clearance — mirrors LibreTranslate.
    commercial_ready = False
    cost_note = "Free public API (rate-limited — be polite; self-host for volume)"

    # --- Locale mapping ---------------------------------------------------

    @staticmethod
    def _map_locale(code: str) -> str:
        """Apertium accepts both ISO 639-1 ('es') and 639-3 ('spa') — pass through."""
        return code

    # --- Endpoint / credentials -------------------------------------------

    def _resolve_endpoint(self) -> str:
        ep = (
            self.options.get("apertium_url")
            or _env_first("APERTIUM_API_URL")
            or APERTIUM_DEFAULT_BASE
        )
        ep = ep.rstrip("/")
        if ep.endswith("/translate"):
            ep = ep[: -len("/translate")]
        return ep

    def _resolve_credentials(self) -> dict:
        # Keyless public API (like LibreTranslate) — a missing key is NOT a
        # configuration error. Include a key only if the deployment needs one.
        return {
            "endpoint": self._resolve_endpoint(),
            "api_key": (
                self.options.get("apertium_api_key")
                or _env_first("APERTIUM_API_KEY")
            ),
        }

    # --- Response parsing (factored out for direct testing) ---------------

    @staticmethod
    def _parse_response(payload: dict) -> str:
        data = (payload or {}).get("responseData") or {}
        translated = data.get("translatedText")
        if translated is None:
            raise ValueError(
                f"Apertium: no translatedText in response: {str(payload)[:200]}"
            )
        return translated

    # --- Backend ----------------------------------------------------------

    async def _translate_texts(
        self,
        texts: list[str],
        src: str,
        tgt: str,
        creds: dict,
    ) -> list[str]:
        # One GET per text. MAX_BATCH=1 means the base loop already hands us a
        # single-element list, but loop defensively so any batch size works.
        langpair = f"{self._map_locale(src)}|{self._map_locale(tgt)}"
        base = creds["endpoint"]
        headers = {"User-Agent": APERTIUM_USER_AGENT}
        out: list[str] = []
        async with aiohttp.ClientSession(headers=headers) as session:
            for text in texts:
                params = {"langpair": langpair, "q": text, "markUnknown": "no"}
                if creds.get("api_key"):
                    params["key"] = creds["api_key"]
                async with session.get(f"{base}/translate", params=params) as resp:
                    if resp.status != 200:
                        body = await resp.text()
                        raise RuntimeError(
                            f"Apertium: HTTP {resp.status} — {body[:200]}"
                        )
                    # content_type=None: APy may send text/json; don't reject it.
                    payload = await resp.json(content_type=None)
                out.append(self._parse_response(payload))
        return out
